client/state/data-layer/wpcom-http/index.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * External dependencies
+ */
+import { get, noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+import { WPCOM_HTTP_REQUEST } from 'state/action-types';
+import { extendAction } from 'state/utils';
+
+/**
+ * Returns the appropriate fetcher in wpcom given the request method
+ *
+ * fetcherMap :: String -> (Params -> Query -> [Body] -> Promise)
+ *
+ * @param {string} method name of HTTP method for request
+ * @returns {Function} the fetcher
+ */
+const fetcherMap = method => get( {
+	GET: wpcom.req.get.bind( wpcom.req ),
+	POST: wpcom.req.post.bind( wpcom.req ),
+}, method, noop );
+
+export const successMeta = data => ( { meta: { dataLayer: { data } } } );
+export const failureMeta = error => ( { meta: { dataLayer: { error } } } );
+export const progressMeta = ( { size: total, loaded } ) => ( { meta: { dataLayer: { progress: { total, loaded } } } } );
+
+const queueRequest = ( { dispatch }, action, next ) => {
+	const {
+		body,
+		formData,
+		method,
+		onSuccess,
+		onFailure,
+		onProgress,
+		path,
+		query,
+	} = action;
+
+	const request = fetcherMap( method.toUpperCase() )(
+		{ path, formData },
```

Comment:
after starting to make this change I disagree with my conclusion. HTTP requests going to WordPress.com APIs have inherently different semantics than those going to random URLs around the internet and they can be classified together as a group.

if we completely generalize then we need to give up `wpcom.req` and start doing things like providing authentication wrappers and locale wrapper.

if we keep them separate then we get that stuff for "free" by means of the fact that it's already built-in to `wpcom.js`

@aduth does this sound reasonable to you?